okhttp-tests/src/test/java/okhttp3/HeadersTest.java

```diff
@@ -434,20 +481,285 @@
+    assertEquals("my\\\"realm", challenges.get(1).realm());
+    expectedAuthParams.clear();
+    expectedAuthParams.put("realm", "my\\\"realm");
+    assertEquals(expectedAuthParams, challenges.get(1).authParams());
+
+    // Comma in the quoted string part.
+    headers = new Headers.Builder()
+        .add("WWW-Authenticate", "Digest,,,, Basic ,,,realm=\"my, realm,\"").build();
+    challenges = HttpHeaders.parseChallenges(headers, "WWW-Authenticate");
+    assertEquals(2, challenges.size());
+    assertEquals("digest", challenges.get(0).scheme());
+    expectedAuthParams.clear();
+    assertEquals(expectedAuthParams, challenges.get(0).authParams());
+    assertEquals("basic", challenges.get(1).scheme());
+    assertEquals("my, realm,", challenges.get(1).realm());
+    expectedAuthParams.clear();
+    expectedAuthParams.put("realm", "my, realm,");
+    assertEquals(expectedAuthParams, challenges.get(1).authParams());
+
+    headers = new Headers.Builder()
+        .add("WWW-Authenticate", ",Digest,,,, Basic ,,,realm=\"my, realm,\"").build();
+    challenges = HttpHeaders.parseChallenges(headers, "WWW-Authenticate");
+    assertEquals(2, challenges.size());
+    assertEquals("digest", challenges.get(0).scheme());
+    expectedAuthParams.clear();
+    assertEquals(expectedAuthParams, challenges.get(0).authParams());
+    assertEquals("basic", challenges.get(1).scheme());
+    assertEquals("my, realm,", challenges.get(1).realm());
+    expectedAuthParams.clear();
+    expectedAuthParams.put("realm", "my, realm,");
+    assertEquals(expectedAuthParams, challenges.get(1).authParams());
+
+    // Unescaped double quote in the quoted string part.
+    headers = new Headers.Builder()
+        .add("WWW-Authenticate", "Digest,,,, Basic ,,,realm=\"my\\\\\\\\\"r\\ealm\"").build();
+    challenges = HttpHeaders.parseChallenges(headers, "WWW-Authenticate");
+    assertEquals(0, challenges.size());
+
+    headers = new Headers.Builder()
+        .add("WWW-Authenticate", "Digest,,,, Basic ,,,realm=\"my\"realm\"").build();
+    challenges = HttpHeaders.parseChallenges(headers, "WWW-Authenticate");
+    assertEquals(0, challenges.size());
+
+    // Double quote in the token part.
+    headers = new Headers.Builder()
+        .add("WWW-Authenticate", "Digest,,,, Basic ,,,realm=my\"realm").build();
+    challenges = HttpHeaders.parseChallenges(headers, "WWW-Authenticate");
+    assertEquals(0, challenges.size());
+
+    // Token68 part instead of auth params.
+    headers = new Headers.Builder()
+        .add("WWW-Authenticate", "Other abc==").build();
+    challenges = HttpHeaders.parseChallenges(headers, "WWW-Authenticate");
+    assertEquals(1, challenges.size());
+    assertEquals("other", challenges.get(0).scheme());
+    assertNull(challenges.get(0).realm());
+    expectedAuthParams.clear();
+    expectedAuthParams.put(null, "abc==");
+    assertEquals(expectedAuthParams, challenges.get(0).authParams());
+
+    // Invalid token68 part and auth params.
```

Comment:
nit: I know the structure is existing, but each of these seem better suited as individual test methods rather than commented blocks.

Reply:
I totally agree, I just did like always with PRs and tried to adapt to the surroundings.
Want me to put all into separate methods, also the existing ones?

Reply:
Oh, I just noticed the test name which makes my tests in there misplaced anyway. :-D